modepy/matrices.py

```diff
@@ -351,61 +352,122 @@ def mass_matrix(
 
-    return modal_mass_matrix
+    test_functions = (
+        test_basis.derivatives(test_derivative_ax)
+        if test_derivative_ax is not None else test_basis.functions
+    )
 
+    vdm = vandermonde(test_basis.functions, nodes)
 
-def nodal_quad_mass_matrix(
+    mapped_nodes = (
+        mapping_function(quadrature.nodes)
+        if mapping_function is not None else quadrature.nodes
+    )
+
+    modal_operator = np.array([
+        test_f(mapped_nodes) * quadrature.weights
+        for test_f in test_functions
+    ])
+
+    return la.solve(vdm.T, modal_operator)
+
+
+def nodal_quadrature_bilinear_form_matrix(
             quadrature: Quadrature,
-            test_functions: Sequence[Callable[[np.ndarray], np.ndarray]],
-            nodes: np.ndarray | None = None,
+            test_basis: Basis,
+            trial_basis: Basis,
+            input_nodes: np.ndarray,
+            output_nodes: np.ndarray | None = None,
+            mapping_function: Callable[[np.ndarray], np.ndarray] | None = None,
+            test_derivative_ax: int | None = None,
+            trial_derivative_ax: int | None = None
         ) -> np.ndarray:
-    r"""Using the *quadrature*, provide a matrix :math:`M` that
-    satisfies:
+    r"""Using *quadrature*, provide a matrix :math:`A` defined as:
 
     .. math::
 
-        \displaystyle (M \boldsymbol u)_i = \sum_j w_j \phi_i(r_j) u_j,
+        \displaystyle A_{ij} = \sum_k \psi_i(r_k) \phi_j(r_k) w_k,
 
-    where :math:`\phi_i` are the (volume) Lagrange basis functions obtained from
-    *test_functions* at *nodes*, :math:`w_i` and :math:`r_i` are the
-    weights and nodes from *qudarature*, and :math:`u_j` are the point
-    values of the trial function at the nodes of *quadrature*.
+    where :math:`\psi_i` are Lagrange basis functions obtained from
+    *test_functions*, :math:`\phi_j` are Lagrange basis functions obtained from
+    *trial_functions*, :math:`r_k` and :math:`w_k` are nodes and weights from
+    *quadrature*. The matrix :math:`A` satisfies
 
-    If *nodes* is not provided, use *quadrature*'s nodes.
+    .. math::
 
-    .. versionadded :: 2024.2
+        \displaystyle (u, \psi_i)_A = \sum_{j} A_{ij} u_j,
+
+    where :math:`u_i` are nodal coefficients.
+
+    An optional *mapping_function* can be supplied to map the *quadrature* nodes
+    evaluated by *test_basis* before constructing the matrix.
+
+    An optional set of *output_nodes* can be supplied if the result should live
+    on a domain different from the domain of *input_nodes*.
+
+    Test and trial function derivatives can be requested by supplying either
+    *test_derivative_ax* or *trial_derivative_ax* as an integer corresponding to
+    an enumerated spatial derivative ax. For example, *test_derivative_ax = 0*
+    would use *test_basis* differentiated with respect to :math:`x` while
+    *trial_derivative_ax = 2* would use *trial_basis* differentiated with
+    respect to :math:`z`.
```

Comment:
Not sure this is worth it, vs. having the user just pass in the derivative function set.